File: zhtml/src/org/zkoss/zhtml/Area.java
/* Area.java

	Purpose:
		
	Description:
		
	History:
		Tue Dec 13 14:58:16     2005, Created by tomyeh

Copyright (C) 2005 Potix Corporation. All Rights Reserved.

{{IS_RIGHT
	This program is distributed under LGPL Version 2.1 in the hope that
	it will be useful, but WITHOUT ANY WARRANTY.
}}IS_RIGHT
*/
package org.zkoss.zhtml;

import org.zkoss.zhtml.impl.AbstractTag;
import org.zkoss.zk.ui.WrongValueException;

/**
 * The AREA tag.
 * 
 * @author tomyeh
 */
public class Area extends AbstractTag {
	public Area() {
		super("area");
	}
	/**
	 * Returns the alt of this area tag.
	 * @since 8.0.3
	 */
	public String getAlt() {
		return (String) getDynamicProperty("alt");
	}

	/**
	 * Sets the alt of this area tag.
	 * @since 8.0.3
	 */
	public void setAlt(String alt) throws WrongValueException {
		setDynamicProperty("alt", alt);
	}
	/**
	 * Returns the coords of this area tag.
	 * @since 8.0.3
	 */
	public String getCoords() {
		return (String) getDynamicProperty("coords");
	}

	/**
	 * Sets the coords of this area tag.
	 * @since 8.0.3
	 */
	public void setCoords(String coords) throws WrongValueException {
		setDynamicProperty("coords", coords);
	}
	/**
	 * Returns the download of this area tag.
	 * @since 8.0.3
	 */
	public String getDownload() {
		return (String) getDynamicProperty("download");
	}

	/**
	 * Sets the download of this area tag.
	 * @since 8.0.3
	 */
	public void setDownload(String download) throws WrongValueException {
		setDynamicProperty("download", download);
	}
	/**
	 * Returns the href of this area tag.
	 * @since 8.0.3
	 */
	public String getHref() {
		return (String) getDynamicProperty("href");
	}

	/**
	 * Sets the href of this area tag.
	 * @since 8.0.3
	 */
	public void setHref(String href) throws WrongValueException {
		setDynamicProperty("href", href);
	}
	/**
	 * Returns the hreflang of this area tag.
	 * @since 8.0.3
	 */
	public String getHreflang() {
		return (String) getDynamicProperty("hreflang");
	}

	/**
	 * Sets the hreflang of this area tag.
	 * @since 8.0.3
	 */
	public void setHreflang(String hreflang) throws WrongValueException {
		setDynamicProperty("hreflang", hreflang);
	}
	/**
	 * Returns the media of this area tag.
	 * @since 8.0.3
	 */
	public String getMedia() {
		return (String) getDynamicProperty("media");
	}

	/**
	 * Sets the media of this area tag.
	 * @since 8.0.3
	 */
	public void setMedia(String media) throws WrongValueException {
		setDynamicProperty("media", media);
	}
	/**
	 * Returns the rel of this area tag.
	 * @since 8.0.3
	 */
	public String getRel() {
		return (String) getDynamicProperty("rel");
	}

	/**
	 * Sets the rel of this area tag.
	 * @since 8.0.3
	 */
	public void setRel(String rel) throws WrongValueException {
		setDynamicProperty("rel", rel);
	}
	/**
	 * Returns the shape of this area tag.
	 * @since 8.0.3
	 */
	public String getShape() {
		return (String) getDynamicProperty("shape");
	}

	/**
	 * Sets the shape of this area tag.
	 * @since 8.0.3
	 */
	public void setShape(String shape) throws WrongValueException {
		setDynamicProperty("shape", shape);
	}
	/**
	 * Returns the target of this area tag.
	 * @since 8.0.3
	 */
	public String getTarget() {
		return (String) getDynamicProperty("target");
	}

	/**
	 * Sets the target of this area tag.
	 * @since 8.0.3
	 */
	public void setTarget(String target) throws WrongValueException {
		setDynamicProperty("target", target);
	}
	/**
	 * Returns the type of this area tag.
	 * @since 8.0.3
	 */
	public String getType() {
		return (String) getDynamicProperty("type");
	}

	/**
	 * Sets the type of this area tag.
	 * @since 8.0.3
	 */
	public void setType(String type) throws WrongValueException {
		setDynamicProperty("type", type);
	}
}

File: zul/src/main/resources/web/js/zul/wgt/Area.ts
/* Area.ts

	Purpose:

	Description:

	History:
		Thu Mar 26 15:54:35     2009, Created by tomyeh

Copyright (C) 2009 Potix Corporation. All Rights Reserved.

This program is distributed under LGPL Version 2.1 in the hope that
it will be useful, but WITHOUT ANY WARRANTY.
*/
/**
 * An area of a {@link Imagemap}.
 */
@zk.WrapClass('zul.wgt.Area')
export class Area extends zk.Widget<HTMLAreaElement> {
	/** @internal */
	_shape?: string;
	/** @internal */
	_coords?: string;

	/**
	 * @returns the shape of this area.
	 * @defaultValue `null` (means rectangle).
	 */
	getShape(): string | undefined {
		return this._shape;
	}

	/**
	 * Sets the shape of this area.
	 *
	 * @param shape - the shape only allows one of
	 * null, "rect", "rectangle", "circle", "circ", "ploygon", and "poly".
	 */
	setShape(shape: string, opts?: Record<string, boolean>): this {
		const o = this._shape;
		this._shape = shape;

		if (o !== shape || opts?.force) {
			var n = this.$n();
			if (n) n.shape = shape || '';
		}

		return this;
	}

	/**
	 * @returns the coordination of this area.
	 */
	getCoords(): string | undefined {
		return this._coords;
	}

	/**
	 * Sets the coords of this area.
	 * Its content depends on {@link getShape}:
	 * <dl>
	 * <dt>circle</dt>
	 * <dd>coords="x,y,r"</dd>
	 * <dt>polygon</dt>
	 * <dd>coords="x1,y1,x2,y2,x3,y3..."<br/>
	 * The polygon is automatically closed, so it is not necessary to repeat
	 * the first coordination.</dd>
	 * <dt>rectangle</dt>
	 * <dd>coords="x1,y1,x2,y2"</dd>
	 * </dl>
	 *
	 * <p>Note: (0, 0) is the upper-left corner.
	 */
	setCoords(coords: string, opts?: Record<string, boolean>): this {
		const o = this._coords;
		this._coords = coords;

		if (o !== coords || opts?.force) {
			// ZK-1892 rename the argument
			var n = this.$n();
			if (n) n.coords = coords || '';
		}

		return this;
	}

	/** @internal */
	override doClick_(evt: zk.Event): void {
		if (zul.wgt.Imagemap._toofast()) return;

		var area = this.id || this.uuid;
		this.parent!.fire('onClick', {area: area}, {ctl: true});
		evt.stop();
	}

	/** @internal */
	override domAttrs_(no?: zk.DomAttrsOptions): string {
		var attr = super.domAttrs_(no)
			+ ' href="javascript:;"', v;
		if (v = this._coords)
			attr += ' coords="' + v + '"';
		if (v = this._shape)
			attr += ' shape="' + v + '"';
		return attr;
	}
}